src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,748 @@
+
+-- | Bitvector add.
+add :: NatRepr w -> BV w -> BV w -> BV w
+add w (BV x) (BV y) = mkBV' w (x+y)
+
+-- | Bitvector subtract.
+sub :: NatRepr w -> BV w -> BV w -> BV w
+sub w (BV x) (BV y) = mkBV' w (x-y)
+
+-- | Bitvector multiply.
+mul :: NatRepr w -> BV w -> BV w -> BV w
+mul w (BV x) (BV y) = mkBV' w (x*y)
+
+-- | Bitvector division (unsigned). Rounds to zero. Division by zero
+-- yields a runtime error.
+uquot :: BV w -> BV w -> BV w
+uquot (BV x) (BV y) = BV (x `quot` y)
+
+-- | Bitvector remainder after division (unsigned), when rounded to
+-- zero. Division by zero yields a runtime error.
+urem :: BV w -> BV w -> BV w
+urem (BV x) (BV y) = BV (x `rem` y)
+
+-- | 'uquot' and 'urem' returned as a pair.
+uquotRem :: BV w -> BV w -> (BV w, BV w)
+uquotRem bv1 bv2 = (uquot bv1 bv2, urem bv1 bv2)
+
+-- | Bitvector division (signed). Rounds to zero. Division by zero
+-- yields a runtime error.
+squot :: NatRepr w -> BV w -> BV w -> BV w
+squot w bv1 bv2 = mkBV' w (x `quot` y)
+  where x = asSigned w bv1
+        y = asSigned w bv2
+
+-- | Bitvector remainder after division (signed), when rounded to
+-- zero. Division by zero yields a runtime error.
+srem :: NatRepr w -> BV w -> BV w -> BV w
+srem w bv1 bv2 = mkBV' w (x `rem` y)
+  where x = asSigned w bv1
+        y = asSigned w bv2
+
+-- | 'squot' and 'srem' returned as a pair.
+squotRem :: NatRepr w -> BV w -> BV w -> (BV w, BV w)
+squotRem w bv1 bv2 = (squot w bv1 bv2, srem w bv1 bv2)
+
+-- | Bitvector division (signed). Rounds to negative infinity. Division
+-- by zero yields a runtime error.
+sdiv :: NatRepr w -> BV w -> BV w -> BV w
+sdiv w bv1 bv2 = mkBV' w (x `div` y)
+  where x = asSigned w bv1
+        y = asSigned w bv2
+
+-- | Bitvector remainder after division (signed), when rounded to
+-- negative infinity. Division by zero yields a runtime error.
+smod :: NatRepr w -> BV w -> BV w -> BV w
+smod w bv1 bv2 = mkBV' w (x `mod` y)
+  where x = asSigned w bv1
+        y = asSigned w bv2
+
+-- | 'sdiv' and 'smod' returned as a pair.
+sdivMod :: NatRepr w -> BV w -> BV w -> (BV w, BV w)
+sdivMod w bv1 bv2 = (sdiv w bv1 bv2, smod w bv1 bv2)
+
+-- | Bitvector absolute value.
+abs :: NatRepr w -> BV w -> BV w
+abs w bv = mkBV' w (Prelude.abs (asSigned w bv))
+
+-- | Bitvector negation.
+negate :: NatRepr w -> BV w -> BV w
+negate w (BV x) = mkBV' w (-x)
+
+-- | Get the sign bit as a 'BV'.
+signBit :: 1 <= w => NatRepr w -> BV w -> BV w
+signBit w bv@(BV _) = lshr bv (natValue w - 1) `and` BV 1
+
+-- | Return 1 if positive, -1 if negative, and 0 if 0.
+signum :: 1 <= w => NatRepr w -> BV w -> BV w
+signum w bv = mkBV' w (Prelude.signum (asSigned w bv))
 
 -- | Signed less than.
-bvLTS :: BitVector w -> BitVector w -> Bool
-bvLTS bv1 bv2 = bvIntegerS bv1 < bvIntegerS bv2
+slt :: NatRepr w -> BV w -> BV w -> Bool
+slt w bv1 bv2 = asSigned w bv1 < asSigned w bv2
+
+-- | Signed less than or equal.
+sle :: NatRepr w -> BV w -> BV w -> Bool
+sle w bv1 bv2 = bv1 == bv2 || slt w bv1 bv2
```

Comment:
Is there a reason to prefer this over just `<=` directly?

Reply:
I think you'd need to do `asSigned bv1 <= asSigned bv2`, which might be better. I'm not entirely sure.

Reply:
Yeah, I'd rather this boiled down into one operation instead of two.  If you don't want do define `sle` directly, you can set `sle w bv1 bv2 = not (slt w bv2 bv1)`